File: SCAPE/src/rotation2twist.py
#The function that translates rotation matrix into twist, from Ma Yi An invitation to 3D Vision

import numpy as np
import math

'''
Input: 3 * 3 Rotation Matrix

Output: 3 * 1 Twist vector
'''
def rotation2twist(matrix_m):
    mid_theta = (np.trace(matrix_m) - 1) / 2
    theta = math.acos(mid_theta)
    t = np.array([[matrix_m[2][1] - matrix_m[1][2]], [matrix_m[0][2] - matrix_m[2][0]], [matrix_m[1][0] - matrix_m[0][1]]]).T
    t = (abs(theta) / (2 * math.sin(theta))) * t
    return t

def test_twist():
    matrix_m = np.array([[1, 0, 0], [0, math.sqrt(2) / 2, -math.sqrt(2) / 2], [0, math.sqrt(2) / 2, math.sqrt(2) / 2]])
    t = rotation2twist(matrix_m)
    print t

if __name__ == '__main__':
    test_twist()

File: SCAPE Code/scape training/rotation2twist.m
function t =  rotation2twist(matrix_m)
%UNTITLED Summary of this function goes here
%   Detailed explanation goes here
    mid_theta = (trace(matrix_m) - 1) / 2;
    theta = acos(mid_theta);
    t = [matrix_m(3, 2) - matrix_m(2, 3); matrix_m(1, 3) - matrix_m(3, 1); matrix_m(2, 1) - matrix_m(1, 2)]';
    t = (abs(theta) / (2 * sin(theta))) * t;
end

